I/O error during migration does not cleanly abort migration

### Describe the bug

A fresh instance on v1.51.2-X01 encountered "org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend." on its first start. On subsequent starts it would fail with "Database has migration lock; cannot run migrations."

### To Reproduce

I did not reproduce it, but expect the following to show the issue:

1. Start Metabase v51.2
2. While it is running migrations, induce network errors, causing a migration to fail
3. Wait for Metabase to shut down
4. Start it again
5. Observe startup failure and "Database has migration lock; cannot run migrations."

### Expected behavior

Upon encountering I/O errors during migrations, either:
* Retry migration
* Release migration lock, so that the next start will be able to successfully apply them

### Logs

First exception:
```
liquibase.exception.DatabaseException: An I/O error occurred while sending to the backend. [Failed SQL: (0) ALTER TABLE query_action ADD CONSTRAINT pk_query_action PRIMARY KEY (action_id)]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:470)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:179)
	at liquibase.executor.AbstractExecutor.execute(AbstractExecutor.java:141)
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1285)
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:755)
	at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:119)
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:68)
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(ChangeLogIterator.java:133)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:122)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Scope.child(Scope.java:252)
	at liquibase.Scope.child(Scope.java:256)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)
	at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(AbstractUpdateCommandStep.java:110)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:108)
	at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:105)
	at liquibase.command.CommandScope.execute(CommandScope.java:217)
	at liquibase.Liquibase.lambda$update$0(Liquibase.java:245)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Liquibase.runInScope(Liquibase.java:1419)
	at liquibase.Liquibase.update(Liquibase.java:234)
	at liquibase.Liquibase.update(Liquibase.java:212)
	at liquibase.Liquibase.update(Liquibase.java:194)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_$fn__45488.invoke(liquibase.clj:371)
	at metabase.db.liquibase$run_in_scope_locked$reify__45484.run(liquibase.clj:336)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:329)
	at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:312)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:360)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:353)
	at metabase.db.setup$migrate_BANG_55501__55502$fn__55503.invoke(setup.clj:84)
	at metabase.db.liquibase$do_with_liquibase45418__45419$f_STAR___45420.invoke(liquibase.clj:135)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invokeStatic(liquibase.clj:138)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invoke(liquibase.clj:126)
	at metabase.db.setup$migrate_BANG_55501__55502.invokeStatic(setup.clj:73)
	at metabase.db.setup$migrate_BANG_55501__55502.doInvoke(setup.clj:55)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invokeStatic(setup.clj:149)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invoke(setup.clj:144)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539$fn__55540.invoke(setup.clj:167)
	at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:238)
	at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:224)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539.invoke(setup.clj:161)
	at metabase.db.setup$setup_db_BANG_55535__55536.invokeStatic(setup.clj:160)
	at metabase.db.setup$setup_db_BANG_55535__55536.invoke(setup.clj:153)
	at metabase.db$setup_db_BANG_$fn__55564.invoke(db.clj:86)
	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:81)
	at metabase.db$setup_db_BANG_.doInvoke(db.clj:68)
	at clojure.lang.RestFn.invoke(RestFn.java:424)
	at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:121)
	at metabase.core$init_BANG__STAR_.invoke(core.clj:102)
	at metabase.core$init_BANG_.invokeStatic(core.clj:181)
	at metabase.core$init_BANG_.invoke(core.clj:176)
	at metabase.core$start_normally.invokeStatic(core.clj:193)
	at metabase.core$start_normally.invoke(core.clj:187)
	at metabase.core$entrypoint.invokeStatic(core.clj:226)
	at metabase.core$entrypoint.doInvoke(core.clj:220)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at clojure.lang.Var.applyTo(Var.java:707)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at metabase.bootstrap.main(Unknown Source)
Caused by: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:398)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:341)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:326)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:302)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:297)
	at com.mchange.v2.c3p0.impl.NewProxyStatement.execute(NewProxyStatement.java:75)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:464)
	... 86 more
Caused by: java.io.EOFException
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:469)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2166)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	... 94 more
```

Then, on the same run:
```
liquibase.exception.LockException: liquibase.exception.DatabaseException: org.postgresql.util.PSQLException: This connection has been closed.
	at liquibase.lockservice.StandardLockService.releaseLock(StandardLockService.java:407)
	at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:139)
	at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:105)
	at liquibase.command.CommandScope.execute(CommandScope.java:217)
	at liquibase.Liquibase.lambda$update$0(Liquibase.java:245)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Liquibase.runInScope(Liquibase.java:1419)
	at liquibase.Liquibase.update(Liquibase.java:234)
	at liquibase.Liquibase.update(Liquibase.java:212)
	at liquibase.Liquibase.update(Liquibase.java:194)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_$fn__45488.invoke(liquibase.clj:371)
	at metabase.db.liquibase$run_in_scope_locked$reify__45484.run(liquibase.clj:336)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:329)
	at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:312)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:360)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:353)
	at metabase.db.setup$migrate_BANG_55501__55502$fn__55503.invoke(setup.clj:84)
	at metabase.db.liquibase$do_with_liquibase45418__45419$f_STAR___45420.invoke(liquibase.clj:135)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invokeStatic(liquibase.clj:138)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invoke(liquibase.clj:126)
	at metabase.db.setup$migrate_BANG_55501__55502.invokeStatic(setup.clj:73)
	at metabase.db.setup$migrate_BANG_55501__55502.doInvoke(setup.clj:55)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invokeStatic(setup.clj:149)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invoke(setup.clj:144)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539$fn__55540.invoke(setup.clj:167)
	at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:238)
	at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:224)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539.invoke(setup.clj:161)
	at metabase.db.setup$setup_db_BANG_55535__55536.invokeStatic(setup.clj:160)
	at metabase.db.setup$setup_db_BANG_55535__55536.invoke(setup.clj:153)
	at metabase.db$setup_db_BANG_$fn__55564.invoke(db.clj:86)
	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:81)
	at metabase.db$setup_db_BANG_.doInvoke(db.clj:68)
	at clojure.lang.RestFn.invoke(RestFn.java:424)
	at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:121)
	at metabase.core$init_BANG__STAR_.invoke(core.clj:102)
	at metabase.core$init_BANG_.invokeStatic(core.clj:181)
	at metabase.core$init_BANG_.invoke(core.clj:176)
	at metabase.core$start_normally.invokeStatic(core.clj:193)
	at metabase.core$start_normally.invoke(core.clj:187)
	at metabase.core$entrypoint.invokeStatic(core.clj:226)
	at metabase.core$entrypoint.doInvoke(core.clj:220)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at clojure.lang.Var.applyTo(Var.java:707)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at metabase.bootstrap.main(Unknown Source)
Caused by: liquibase.exception.DatabaseException: org.postgresql.util.PSQLException: This connection has been closed.
	at liquibase.database.jvm.JdbcConnection.rollback(JdbcConnection.java:464)
	at liquibase.database.AbstractJdbcDatabase.rollback(AbstractJdbcDatabase.java:1203)
	at liquibase.database.core.PostgresDatabase.rollback(PostgresDatabase.java:401)
	at liquibase.lockservice.StandardLockService.releaseLock(StandardLockService.java:370)
	... 61 more
Caused by: org.postgresql.util.PSQLException: This connection has been closed.
	at org.postgresql.jdbc.PgConnection.checkClosed(PgConnection.java:1009)
	at org.postgresql.jdbc.PgConnection.getAutoCommit(PgConnection.java:969)
	at com.mchange.v2.c3p0.impl.NewProxyConnection.getAutoCommit(NewProxyConnection.java:1227)
	at liquibase.database.jvm.JdbcConnection.rollback(JdbcConnection.java:460)
	... 64 more
```

On next start:
```
liquibase.exception.LockException: Database has migration lock; cannot run migrations. You can force-release these locks by running `java -jar metabase.jar migrate release-locks`.
	at metabase.db.liquibase$wait_for_migration_lock$fn__45449.invoke(liquibase.clj:233)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:174)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:181)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:181)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:181)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:181)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.util.jvm$do_with_auto_retries.invokeStatic(jvm.clj:181)
	at metabase.util.jvm$do_with_auto_retries.invoke(jvm.clj:164)
	at metabase.db.liquibase$wait_for_migration_lock.invokeStatic(liquibase.clj:227)
	at metabase.db.liquibase$wait_for_migration_lock.invoke(liquibase.clj:222)
	at metabase.db.liquibase$run_in_scope_locked$reify__45484.run(liquibase.clj:333)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:329)
	at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:312)
	at metabase.db.liquibase$consolidate_liquibase_changesets_BANG_45562__45563.invokeStatic(liquibase.clj:464)
	at metabase.db.liquibase$consolidate_liquibase_changesets_BANG_45562__45563.invoke(liquibase.clj:443)
	at metabase.db.setup$migrate_BANG_55501__55502$fn__55503.invoke(setup.clj:80)
	at metabase.db.liquibase$do_with_liquibase45418__45419$f_STAR___45420.invoke(liquibase.clj:135)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invokeStatic(liquibase.clj:138)
	at metabase.db.liquibase$do_with_liquibase45418__45419.invoke(liquibase.clj:126)
	at metabase.db.setup$migrate_BANG_55501__55502.invokeStatic(setup.clj:73)
	at metabase.db.setup$migrate_BANG_55501__55502.doInvoke(setup.clj:55)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invokeStatic(setup.clj:149)
	at metabase.db.setup$run_schema_migrations_BANG_55528__55529.invoke(setup.clj:144)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539$fn__55540.invoke(setup.clj:167)
	at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:238)
	at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:224)
	at metabase.db.setup$setup_db_BANG_55535__55536$fn__55539.invoke(setup.clj:161)
	at metabase.db.setup$setup_db_BANG_55535__55536.invokeStatic(setup.clj:160)
	at metabase.db.setup$setup_db_BANG_55535__55536.invoke(setup.clj:153)
	at metabase.db$setup_db_BANG_$fn__55564.invoke(db.clj:86)
	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:81)
	at metabase.db$setup_db_BANG_.doInvoke(db.clj:68)
	at clojure.lang.RestFn.invoke(RestFn.java:424)
	at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:121)
	at metabase.core$init_BANG__STAR_.invoke(core.clj:102)
	at metabase.core$init_BANG_.invokeStatic(core.clj:181)
	at metabase.core$init_BANG_.invoke(core.clj:176)
	at metabase.core$start_normally.invokeStatic(core.clj:193)
	at metabase.core$start_normally.invoke(core.clj:187)
	at metabase.core$entrypoint.invokeStatic(core.clj:226)
	at metabase.core$entrypoint.doInvoke(core.clj:220)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at clojure.lang.Var.applyTo(Var.java:707)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at metabase.bootstrap.main(Unknown Source)
```

### Information about your Metabase installation

```JSON
Metabase v1.51.2 on Linux
```

### Severity

can cause downtime, but easy workaround

### Additional context

https://metaboat.slack.com/archives/C013VC33N95/p1732025675282919

Similar to: https://github.com/metabase/metabase/issues/37866

Different issue, not the problem I ran into, but similar cause: https://github.com/metabase/metabase/issues/30360

--- a/deps.edn
+++ b/deps.edn
@@ -40,6 +40,7 @@
   com.fasterxml.woodstox/woodstox-core      {:mvn/version "7.1.0"}              ; trans dep of commons-codec
   ; Detect the charset in uploaded CSV files
   com.github.albfernandez/juniversalchardet {:mvn/version "2.5.0"}
+  com.github.blagerweij/liquibase-sessionlock {:mvn/version "1.6.9"}            ; session lock for liquibase migrations
   com.github.jknack/handlebars              ^:antq/exclude                      ; 4.4.0 requires java 17+
                                             {:mvn/version "4.3.1"}              ; templating engine
   com.github.seancorfield/honeysql          {:mvn/version "2.6.1270"}           ; Honey SQL 2. SQL generation from Clojure data maps

--- a/resources/migrations/000_legacy_migrations.yaml
+++ b/resources/migrations/000_legacy_migrations.yaml
@@ -5829,16 +5829,14 @@ databaseChangeLog:
   - changeSet:
         id: 109
         author: camsaul
-        comment: Added 0.34.2
+        comment: Added 0.34.2, but not necessary anymore with the liquibase session lock
         failOnError: false
         preConditions:
           - onFail: MARK_RAN
           - onSqlOutput: FAIL
           - or:
               - dbms:
-                    type: mysql
-              - dbms:
-                    type: mariadb
+                    type: none
         changes:
           - sql:
                 sql: ALTER TABLE `DATABASECHANGELOGLOCK` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;

--- a/src/metabase/db/liquibase.clj
+++ b/src/metabase/db/liquibase.clj
@@ -4,6 +4,7 @@
    [clojure.java.jdbc :as jdbc]
    [clojure.string :as str]
    [metabase.config :as config]
+   [metabase.db.connection :as mdb.connection]
    [metabase.db.custom-migrations]
    [metabase.db.liquibase.h2 :as liquibase.h2]
    [metabase.db.liquibase.mysql :as liquibase.mysql]
@@ -204,15 +205,19 @@
 (defn force-release-locks!
   "(Attempt to) force release Liquibase migration locks."
   [^Liquibase liquibase]
-  (.forceReleaseLocks liquibase))
+  (if-not (= (mdb.connection/db-type) :h2)
+    (throw (ex-info "Database %s does not support forcing migration locks. The lock is always automatically removed when the process performing the migration completes or fails."
+                    {:db-type (mdb.connection/db-type)}))
+    (.forceReleaseLocks liquibase)))
 
 (defn release-lock-if-needed!
-  "Attempts to release the liquibase lock if present. Logs but does not bubble up the exception if one occurs as it's
+  "Attempts to release the liquibase lock on h2 if present. Logs but does not bubble up the exception if one occurs as it's
   intended to be used when a failure has occurred and bubbling up this exception would hide the real exception."
   [^Liquibase liquibase]
   (when (migration-lock-exists? liquibase)
     (try
-      (force-release-locks! liquibase)
+      (when (= (mdb.connection/db-type) :h2)
+        (force-release-locks! liquibase))
       (catch Exception e
         (log/error e "Unable to release the Liquibase lock")))))
 

--- a/test/metabase/db/force_migration_test.clj
+++ b/test/metabase/db/force_migration_test.clj
@@ -28,8 +28,8 @@
               {:name "Egypt"
                :period "3100BC"}))
 
-;; This test tests 4 things:
-;; - It forces release the locks if there is one
+;; This test tests 4 (on h2) or 3 (on others) things:
+;; - It forces release the locks if there is one (only on h2, others use a session lock that can't be forced)
 ;; - When running migrations, when a migration fails, ignores it and move on to the next migration
 ;; - All custom migrations are executed in a transaction, if it fails, nothing should be commited
 ;; - All migrations are executed in the order it's defined
@@ -44,7 +44,8 @@
                                 (#'liquibase/liquibase-connection)
                                 (#'liquibase/database))
               lock-service (.getLockService (LockServiceFactory/getInstance) database)]
-          (.acquireLock lock-service)
+          (when (= driver/*driver* :h2)
+            (.acquireLock lock-service))
           (db.setup/migrate! data-source :force)
 
           (testing "Make sure the migrations that intended to succeed are succeed"

--- a/test/metabase/db/general_schema_tests.clj
+++ b/test/metabase/db/general_schema_tests.clj
@@ -10,11 +10,10 @@
 (deftest ^:parallel no-tiny-int-columns
   (when (= (mdb/db-type) :mysql)
     (testing "All boolean columns in mysql, mariadb should be bit(1)"
-      (is (= [{:table_name "DATABASECHANGELOGLOCK" :column_name "LOCKED"}] ;; outlier because this is liquibase's table
-             (t2/query
-              (format "SELECT table_name, column_name FROM information_schema.columns WHERE data_type LIKE 'tinyint%%' AND table_schema = '%s';"
-                      (with-open [conn (-> (mdb/app-db) .getConnection)]
-                        (.getCatalog conn)))))))))
+      (is (= [] (t2/query
+                 (format "SELECT table_name, column_name FROM information_schema.columns WHERE data_type LIKE 'tinyint%%' AND table_schema = '%s';"
+                         (with-open [conn (-> (mdb/app-db) .getConnection)]
+                           (.getCatalog conn)))))))))
 
 (deftest ^:parallel fks-are-indexed-test
   (when (= (mdb/db-type) :postgres)

--- a/test/metabase/db/liquibase_test.clj
+++ b/test/metabase/db/liquibase_test.clj
@@ -4,6 +4,7 @@
    [clojure.set :as set]
    [clojure.string :as str]
    [clojure.test :refer :all]
+   [metabase.db :as mdb]
    [metabase.db.liquibase :as liquibase]
    [metabase.db.test-util :as mdb.test-util]
    [metabase.driver :as driver]
@@ -12,7 +13,10 @@
    [metabase.test :as mt]
    [metabase.util.yaml :as u.yaml]
    [next.jdbc :as next.jdbc]
-   [toucan2.core :as t2]))
+   [toucan2.core :as t2])
+  (:import
+   (liquibase Liquibase)
+   (liquibase.lockservice LockServiceFactory)))
 
 (set! *warn-on-reflection* true)
 
@@ -128,3 +132,24 @@
             (deliver released true)
             (testing "The lock was released before the migration finished"
               (is (not @locked?)))))))))
+
+(deftest auto-release-session-lock-test
+  (mt/test-drivers #{:mysql :postgres}
+    (testing "Session lock is released on conn close"
+      ;; Session lock provided automatically by the com.github.blagerweij/liquibase-sessionlock dependency
+      (mt/with-temp-empty-app-db [_conn driver/*driver*]
+        (let [;; use data-source so with-liquibase opens and closes the conn itself
+              data-source (mdb/data-source)
+              lock        (fn [^Liquibase liquibase]
+                            (->> liquibase
+                                 .getDatabase
+                                 (.getLockService (LockServiceFactory/getInstance))
+                                 .acquireLock))]
+          (liquibase/with-liquibase [liquibase1 data-source]
+            (is (lock liquibase1) "Can initially acquire session lock")
+            (is (lock liquibase1) "Can require acquire session lock on same liquibase")
+            (liquibase/with-liquibase [liquibase2 data-source]
+              (is (not (lock liquibase2)) "Cannot acquire session lock on a different liquibase while it is taken")))
+          (liquibase/with-liquibase [liquibase3 data-source]
+            ;; This will fail if the com.github.blagerweij/liquibase-sessionlock dep is not present
+            (is (lock liquibase3) "Can acquire session lock when conn closed without lock release")))))))

--- a/test/metabase/db/liquibase_test.clj
+++ b/test/metabase/db/liquibase_test.clj
@@ -53,9 +53,7 @@
   "Read a liquibase migration file and returns all the migration id that is applied to `db-type`.
   Ids are orderer in the order it's defined in migration file."
   [file-path db-type]
-  (let [content (u.yaml/from-file (io/resource file-path))
-        ;; 109 is a DATABASECHANGELOGLOCK alter that isn't used anymore and does not fail on error
-        migration-lock-changeset? (fn [id] (and (#{:mysql :mariadb} db-type) (#{"109"} id)))]
+  (let [content (u.yaml/from-file (io/resource file-path))]
     (->> (:databaseChangeLog content)
          ;; if the changelog has filter by dbms, remove the ones that doens't apply for the current db-type
          (remove (fn [{{:keys [dbms]} :changeSet}] (and (not (str/blank? dbms))
@@ -63,7 +61,6 @@
          ;; remove ignored changeSets
          (remove #(get-in % [:changeSet :ignore]))
          (map #(str (get-in % [:changeSet :id])))
-         (remove migration-lock-changeset?)
          (remove str/blank?))))
 
 (deftest consolidate-liquibase-changesets-test

--- a/test/metabase/db/liquibase_test.clj
+++ b/test/metabase/db/liquibase_test.clj
@@ -110,7 +110,7 @@
               (is (= :done (liquibase/wait-for-all-locks sleep-ms timeout-ms))))))))))
 
 (deftest release-all-locks-if-needed!-test
-  (mt/test-drivers #{:h2 :mysql :postgres}
+  (mt/test-drivers #{:h2}
     (mt/with-temp-empty-app-db [conn driver/*driver*]
       (liquibase/with-liquibase [liquibase conn]
         (testing "When we release the locks from outside the migration...\n"

--- a/test/metabase/db/liquibase_test.clj
+++ b/test/metabase/db/liquibase_test.clj
@@ -53,7 +53,9 @@
   "Read a liquibase migration file and returns all the migration id that is applied to `db-type`.
   Ids are orderer in the order it's defined in migration file."
   [file-path db-type]
-  (let [content (u.yaml/from-file (io/resource file-path))]
+  (let [content (u.yaml/from-file (io/resource file-path))
+        ;; 109 is a DATABASECHANGELOGLOCK alter that isn't used anymore and does not fail on error
+        migration-lock-changeset? (fn [id] (and (#{:mysql :mariadb} db-type) (#{"109"} id)))]
     (->> (:databaseChangeLog content)
          ;; if the changelog has filter by dbms, remove the ones that doens't apply for the current db-type
          (remove (fn [{{:keys [dbms]} :changeSet}] (and (not (str/blank? dbms))
@@ -61,6 +63,7 @@
          ;; remove ignored changeSets
          (remove #(get-in % [:changeSet :ignore]))
          (map #(str (get-in % [:changeSet :id])))
+         (remove migration-lock-changeset?)
          (remove str/blank?))))
 
 (deftest consolidate-liquibase-changesets-test

